### DISCUSSION TITLE
When content was extracted from content stream using doc.xref_stream(xref).decode("ISO-8859-1"), the output extracted was in uninterpretable format.

### DISCUSSION THREAD

**Discussion by meghanaviyyapu** [pymupdf/PyMuPDF#1366] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/36301306/140264903-edc90dc1-b1e8-4147-bbd5-d253bb640a07.png)
Can you please let me know if there is a way to extract text from stream in an interpretable format?

**Comment by meghanaviyyapu** [comment]
[table.pdf](https://github.com/pymupdf/PyMuPDF/files/7472625/table.pdf)
I have considered this PDF as input

**Reply by JorjMcKie** [reply]
>  Can you please let me know if there is a way to extract text from stream in an interpretable format?

Normally there is no easy way to do this. Depends on the font used. The hex numbers <0048> etc. represent glyph ids (not unicode numbers), which the font - via its binary fontfile - converts into a series of drawing commands to form the visual character appearance.
The translation from some character's unicode to glyph id 0x48 is done by that fontfile.
On exception cases like the Base 14 fonts, the unicode may be directly used as the glyph id,

Apart from all that, your excerpt looks pretty damaged, though ...

**Reply by JorjMcKie** [reply]
If you use `page.get_text(...)` all that backtranslation will be done for you, btw.

**Reply by meghanaviyyapu** [reply]
Yes, page.get_text() gives the text but I need to identify text in the stream. I want to identify text in the content stream and generate MCIDs to tag a PDF for accessibility. To do that I need to know the text in content stream. Can you please let me know if this is possible?

**Reply by JorjMcKie** [reply]
>  Can you please let me know if this is possible?

With PyMuPDF not directly / programmatically as I wrote.
You could do some heuristics, like extracting the text so that the fontnames are revealed, then match those fontnames to to the font reference names ("F1", "F2", ...) in 
```python
>>> pprint(page.get_fonts(True))
[(7, 'ttf', 'TrueType', 'ABCDEE+Verdana', 'F1', 'WinAnsiEncoding', 0),
 (11, 'ttf', 'TrueType', 'ABCDEE+Verdana,Bold', 'F2', 'WinAnsiEncoding', 0),
 (37, 'ttf', 'Type0', 'ABCDEE+Verdana', 'F7', 'Identity-H', 0),
 (27, 'n/a', 'TrueType', 'Times New Roman', 'F5', 'WinAnsiEncoding', 0)]
>>> 
```
You are a bit lucky with your file, because most of the text in the `/Contents` is in fact clearly readable (your example hex print out does not occurs on the 2 pages you sent me).
Then locate clearly readable text before and after an unreadable piece and interpret back was has been written there, etc.

Can't think of anything better now.
BTW. there is an abuse of PDF specification in the file on page 2:
The same fontname 'ABCDEE+Verdana' for two different fonts (look at the xrefs 7, 37) - that should not be done:  at least the subsetting characters ("ABCDEE") shall be different.

**Reply by meghanaviyyapu** [reply]
[table.pdf](https://github.com/pymupdf/PyMuPDF/files/7473504/table.pdf)

The /Contents when extracted from this PDF gave results like this
![image](https://user-images.githubusercontent.com/36301306/140285660-09bb7c8b-5ed1-4535-9c09-dd0256984b73.png)
Can you give suggestions to decode the content from the stream.

**Reply by JorjMcKie** [reply]
ok, whatever, that was an aside remark and does not change my answer

**Reply by meghanaviyyapu** [reply]
Okay thanks for the answer.
If the text is completely uninterpretable, can you please let me know if a new content stream can be created

**Reply by JorjMcKie** [reply]
> if a new content stream can be created

No, impossible.

**Reply by meghanaviyyapu** [reply]
Okay thank you.

**Comment by JorjMcKie** [comment, marked as answer]
This is a typical Discussions item. Let me convert it first.
